maskrcnn_benchmark/utils/comm.py

```diff
@@ -64,94 +58,73 @@ def _send_and_wait(r):
     Returns:
-        result (list): a list with as many elements as there are processes,
-            where each element i in the list corresponds to the data that was
-            gathered from the process of rank i.
+        list[data]: list of data gathered from each rank
     """
-    # strategy: the main process creates a temporary directory, and communicates
-    # the location of the temporary directory to all other processes.
-    # each process will then serialize the data to the folder defined by
-    # the main process, and then the main process reads all of the serialized
-    # files and returns them in a list
-    if not torch.distributed.is_available():
-        return [data]
-    if not torch.distributed.is_initialized():
+    world_size = get_world_size()
+    if world_size == 1:
         return [data]
-    synchronize()
-    # get rank of the current process
-    rank = torch.distributed.get_rank()
-
-    # the data to communicate should be small
-    data_to_communicate = torch.empty(256, dtype=torch.uint8, device="cuda")
-    if rank == 0:
-        # manually creates a temporary directory, that needs to be cleaned
-        # afterwards
-        tmp_dir = tempfile.mkdtemp()
-        _encode(data_to_communicate, tmp_dir)
-
-    synchronize()
-    # the main process (rank=0) communicates the data to all processes
-    torch.distributed.broadcast(data_to_communicate, 0)
-
-    # get the data that was communicated
-    tmp_dir = _decode(data_to_communicate)
-
-    # each process serializes to a different file
-    file_template = "file{}.pth"
-    tmp_file = os.path.join(tmp_dir, file_template.format(rank))
-    torch.save(data, tmp_file)
-
-    # synchronize before loading the data
-    synchronize()
-
-    # only the master process returns the data
-    if rank == 0:
-        data_list = []
-        world_size = torch.distributed.get_world_size()
-        for r in range(world_size):
-            file_path = os.path.join(tmp_dir, file_template.format(r))
-            d = torch.load(file_path)
-            data_list.append(d)
-            # cleanup
-            os.remove(file_path)
-        # cleanup
-        os.rmdir(tmp_dir)
-        return data_list
+
+    # serialized to a Tensor
+    buffer = pickle.dumps(data)
+    storage = torch.ByteStorage.from_buffer(buffer)
+    tensor = torch.ByteTensor(storage).to("cuda")
+
+    # obtain Tensor size of each rank
+    local_size = torch.IntTensor([tensor.numel()]).to("cuda")
+    size_list = [torch.IntTensor([0]).to("cuda") for _ in range(world_size)]
+    dist.all_gather(size_list, local_size)
+    size_list = [int(size.item()) for size in size_list]
+    max_size = max(size_list)
+
+    # receiving Tensor from all ranks
+    # we pad the tensor because torch all_gather does not support
+    # gathering tensors of different shapes
+    tensor_list = []
+    for _ in size_list:
+        tensor_list.append(torch.ByteTensor(size=(max_size,)).to("cuda"))
```

Comment:
nit: it would be good to use the new API for this:

```python
device=torch.device("cuda")
...
torch.empty((max_size,), dtype=torch.uint8, device=device
```